fdbserver/workloads/ClogTlog.actor.cpp

```diff
@@ -0,0 +1,184 @@
+			const auto& ip = process->address.ip;
+			if (process->startingClass != ProcessClass::TesterClass) {
+				ips.push_back(ip);
+			}
+		}
+		ASSERT(ips.size() > 0);
+
+		// Find all primary tlogs
+		std::vector<NetworkAddress> logs; // all primary logs except CC
+		for (int i = 0; i < dbInfo->get().logSystemConfig.tLogs.size(); i++) {
+			const auto& tlogset = dbInfo->get().logSystemConfig.tLogs[i];
+			if (!tlogset.isLocal)
+				continue;
+			for (const auto& log : tlogset.tLogs) {
+				const NetworkAddress& addr = log.interf().address();
+				if (cc != addr.ip) {
+					if (!tlog.present()) {
+						tlog = addr;
+					}
+					logs.push_back(addr);
+				}
+			}
+		}
+		ASSERT(logs.size() > 0 && tlog.present());
+
+		// clog pairs
+		for (const auto& ip : ips) {
+			if (ip != tlog.get().ip && ip != cc) {
+				g_simulator->clogPair(ip, tlog.get().ip, seconds);
+				g_simulator->clogPair(tlog.get().ip, ip, seconds);
+				cloggedPairs.emplace_back(ip, tlog.get().ip);
+				cloggedPairs.emplace_back(tlog.get().ip, ip);
+			}
+		}
+		clogged = true;
+	}
+
+	void unclogAll() {
+		// unclog previously clogged connections
+		for (const auto& pair : cloggedPairs) {
+			g_simulator->unclogPair(pair.first, pair.second);
+		}
+		cloggedPairs.clear();
+	}
+
+	ACTOR static Future<Void> excludeFailedLog(ClogTlogWorkload* self, Database cx) {
+		state Future<Void> timeout = delay(30);
+
+		loop choose {
+			when(wait(self->dbInfo->onChange())) {
+				if (self->dbInfo->get().recoveryState >= RecoveryState::ACCEPTING_COMMITS) {
+					return Void();
+				}
+				timeout = delay(30);
+			}
+			when(wait(timeout)) {
+				// recovery state hasn't changed in 30s, exclude the failed tlog
+				std::vector<StringRef> tokens;
+				state Optional<ConfigureAutoResult> conf;
+
+				CODE_PROBE(true, "Exclude failed tlog");
+				TraceEvent("ExcludeFailedLog")
+				    .detail("TLog", self->tlog.get())
+				    .detail("RecoveryState", self->dbInfo->get().recoveryState);
+				tokens.push_back(StringRef("exclude=" + formatIpPort(self->tlog.get().ip, self->tlog.get().port)));
+				ConfigurationResult r =
+				    wait(ManagementAPI::changeConfig(cx.getReference(), tokens, conf, /*force=*/true));
+				TraceEvent("ExcludeFailedLog")
+				    .detail("Result", r)
+				    .detail("ConfigIsValid", conf.present() && conf.get().isValid());
+				return Void();
+			}
+		}
+	}
+
+	ACTOR Future<Void> clogClient(ClogTlogWorkload* self, Database cx) {
+		// Let cycle workload issue some transactions.
+		wait(delay(20.0));
+
+		while (self->dbInfo->get().recoveryState < RecoveryState::FULLY_RECOVERED) {
+			wait(self->dbInfo->onChange());
+		}
+
+		double startTime = now();
+		state double workloadEnd = now() + self->testDuration;
+		TraceEvent("ClogTlog").detail("StartTime", startTime).detail("EndTime", workloadEnd);
+
+		// Clog and wait for recovery to happen
+		self->clogTlog(workloadEnd - now());
+		while (self->dbInfo->get().recoveryState == RecoveryState::FULLY_RECOVERED) {
```

Comment:
Should this stage be `initializing transaction servers`? It shouldn't reach to `FULLY_RECOVERED` isn't it?

Reply:
At line 149, we ensure it's `FULLY_RECOVERED`, so here we wait for the recovery to happen.

Reply:
What about line 158 which partitions a tlog from the rest of the cluster? If that triggers, will we never reach to FULLY_RECOVERED here?

Reply:
158 does the partition, but there is no time lapse (from 149) to this line, so the `FULLY_RECOVERED` state hasn't changed yet. The recovery will happen in 20+s when proxies can't make progress.

Reply:
I finally see what you are doing here. You are waiting it to `NOT` be `FULLY_RECOVERED`. Make sense now. Thanks for explaining.